api/agent.go

```diff
@@ -322,15 +333,60 @@ func NewControllerWrapper(config *loadtest.Config, controllerConfig interface{},
 		}
 	}
 
+	var creds []string
+	f, err := os.Open(config.UsersConfiguration.UsersFilePath)
+	if err == nil {
+		// The file is an optional parameter which may not be present.
+		// So we read it only if it opens successfully.
+		defer f.Close()
+		scanner := bufio.NewScanner(f)
+		for scanner.Scan() {
+			creds = append(creds, scanner.Text())
+		}
+		if err := scanner.Err(); err != nil {
+			return nil, fmt.Errorf("failed to read from %s: %w", f.Name(), err)
+		}
+	} else if !os.IsNotExist(err) {
+		// If the file exists, but failed to open, then probably it is worthwhile
+		// logging it.
+		mlog.Warn("Failed to open UsersFilePath. Continuing with default signup", mlog.Err(err))
+	}
+
 	return func(id int, status chan<- control.UserStatus) (control.UserController, error) {
 		id += userOffset
 
+		var username, password, email string
+
+		if len(creds) > 0 {
+			// We do not allow to mix user-defined and auto signup modes.
+			// If the user specifies a custom login file, then it should contain
+			// all the users that is expected to login during a load-test.
+			if id >= len(creds) {
+				return nil, errors.New("exceeded user list limit")
+			}
+			// Emails and passwords are separated by space.
+			split := strings.Split(creds[id], " ")
+			if len(split) < 2 {
+				return nil, errors.New("user credential does not have space in between")
+			}
+			email = split[0]
+			password = split[1]
+			// Quick and dirty hack to extract username from email.
+			// This is not terribly important to be correct.
+			username = strings.Split(email, "@")[0]
+			username = strings.Replace(username, "+", "-", -1)
+		} else {
+			username = fmt.Sprintf("%s-%d", namePrefix, id)
+			email = fmt.Sprintf("%s-%d@example.com", namePrefix, id)
+			password = "testPass123$"
+		}
```

Comment:
I think all this logic would be better inside a dedicated function that takes the `UsersFilePath` and returns a list of either `model.User` or a smaller ad-hoc structure with the three values we need (username, email and password).

Reply:
I tried to put the entire logic in a separate function, but for the normal case, it would unnecessarily allocate a big slice. And there is the case of the custom offset also. Which might increase the slice further. So kept the user creation part as-is. But open to refactoring this too. Let me know.

Reply:
Are we talking about the `creds` slice? If the path is not defined/empty we shouldn't be allocating anything.

Reply:
But you wanted a slice of users for both cases right? And then the inner function would just pick up the user from that index.

Reply:
I was actually only thinking about this case. Do the whole parsing before we run the load-test and return a slice of [username,email,password] tuples/structs.

Reply:
Ok, understood.